kotest/kotest
# Gradle Multiplatform plugin: set Kotest Compiler dependency version as a constant, configurable in the extension

Update: I'm breaking this PR up into smaller PRs

* #3167 
* #3168
* #3163

---

Updates the Multiplatform Gradle plugin to have a `const val` for the compiler version. This is automatically set by a Gradle task. The version can be manually overridden in the Kotest extension.

> Manual version override example: 
> 
> ```kotlin
> //build.gradle.kts
> plugins {
>    id("io.kotest.multiplatform") version "5.4.0"
> }
> 
> kotest {
>   kotestCompilerVersion.set("5.4.2")
> }
> ```


I also added additional tests using Gradle TestKit for this.

This change is required because the method of determining the version from the project dependencies was not robust. It could break when the version wasn't explicitly set (e.g. it's determined by a BOM, or it's a project dependency or a direct jar), and the order of the Gradle plugins could make a difference (e.g. if the Kotest plugin is applied before the Kotlin Mutliplatform plugin, then no dependencies will be caught).

resolves #3153 

I tried really hard to find a way to automatically determine the version from the dependencies, or from the Gradle buildscript plugins classpath, but this was far too difficult, and almost impossible to do in a way that covered all edgecases. Some sort of fallback `const val` would be necessary, and if there's a fallback then we might as well not bother with trying to figure it out in a complicated way.


#### Summary

- move KotestPluginExtension into different file
- move test-project out of resource directory (so IntelliJ doesn't try to resolve the compiled files in test-project/build
- test-project adjustments to improve manually opening it in a different IntelliJ instance
- apply kotlin-dsl plugin (DSL matches the .kts files better, and aligns Gradle/Kotlin versions)
- replace test-prep in Gradle script with running `build -x check`
- enable caching in test project
- fix gradlew script resolution on Windows
- fix test.xml line comparison on Windows
